### PR TITLE
[AVR] Wrap out-of-bounds relative jumps

### PR DESCRIPTION
**Pull request by Patryk27**
This commit improves the relative jumps, so that we are able to emit `rjmp` that wraps around the memory boundary.

# Motivation

Over https://github.com/Rahix/avr-hal/pull/585 it was found out that the AVR codegen sometimes crashes with the `out of range branch target` message - this turned out to be a fallout from https://github.com/llvm/llvm-project/pull/106722 related to the fact that linkers are smarter than they seem!

The issue stems from the `rjmp` instruction - even though it can only encode offsets in the +-4 kB range, it can be used to jump into further memory addresses by relying on the fact that AVR memory wraps around the boundaries. That is, for a CPU with 8 kB of flash, `rjmp 5000` (illegal) can be actually encoded as `rjmp -3192` (legal).

Linkers apply this trick automatically, but since over https://github.com/llvm/llvm-project/pull/106722 we've decided to go with the `let's avoid emitting relocations` route...

We would've caught this, but since the `branch-relaxation-long.ll` test generated only the textual assembly, it never went near the assertion - I've changed the test to use `-filetype=obj | llvm-objdump -dr` now (which causes it to crash on the main branch, but succeeds with my changes here).

### CODE REVIEW

**Comment by Patryk27** [comment]
cc @benshi001

**Comment by benshi001** [comment]
I think this change is so aggressive, just for special devices with 8KB but without long range JMP/CALL.

I suggest this way:
1. Introduce a new target feature, something like "RjmpWrap" (or a better name??)
2. Add this target feature to these special devices.
3. And then we perform force relocation with this new target feature specified.

**Review by benshi001 — COMMENTED** [review]
I think this change is so aggressive, just for some special devices with 8KB but without long range JMP/CALL.

I suggest this way:

1. Introduce a new target feature, something like "RjmpWrap" (or a better name??)
2. Add this target feature to these special devices.
3. And then we only perform force relocation with this new target feature specified.

**Comment by benshi001** [comment]
> Hmm, so we'd like to diverge from avr-gcc's behavior, right?

Unless there is a very strong reason force us to follow avr-gcc.

**Comment by benshi001** [comment]
> Hmm, so we'd like to diverge from avr-gcc's behavior, right?

Is there a strong reason for us to strictly follow avr-gcc? My key concern is that the AVR backend has been stable, we need not do aggressive functional changes.

**Comment by benshi001** [comment]
But I also think supporting the flash space wrapping is necessary. If you have no time for my suggested solution, I will try it later this month.

**Comment by Patryk27** [comment]
> Is there a strong reason for us to strictly follow avr-gcc?

No, I don't think so - I'm up for avoiding relocations as well 🙂 

> If you have no time for my suggested solution, I will try it later this month.

I can do that, no problem - I'm just reflecting on the approach:

- Option A: Create a new feature (like `RjmpWrap`) and enable it for the smaller (8 kB) microcontrollers,
- Option B: Teach LLVM flash sizes of _all_ of the AVRs and implement logic that automatically applies the memory wrapping trick.

Option A is simpler to implement, while option B (as described in the `Alternatives` approach at the top) is potentially more versatile (we'll be able to apply the wrapping trick even for larger microcontrollers, if we happen to generate a branch large enough). Note that for larger uCs this trick is just a program-size optimization, not a requirement.

I'm leaning towards option B, since it feels less hacky to me, but I don't have any strong preference here - care to choose?

**Comment by benshi001** [comment]
> > Is there a strong reason for us to strictly follow avr-gcc?
> 
> No, I don't think so - I'm up for avoiding relocations as well 🙂
> 
> > If you have no time for my suggested solution, I will try it later this month.
> 
> I can do that, no problem - I'm just reflecting on the approach:
> 
> * Option A: Create a new feature (like `RjmpWrap`) and enable it for the smaller (8 kB) microcontrollers,
> * Option B: Teach LLVM flash sizes of _all_ of the AVRs and implement logic that automatically applies the memory wrapping trick.
> 
> Option A is simpler to implement, while option B (as described in the `Alternatives` approach at the top) is potentially more versatile (we'll be able to apply the wrapping trick even for larger microcontrollers, if we happen to generate a branch large enough). Note that for larger uCs this trick is just a program-size optimization, not a requirement.
> 
> I'm leaning towards option B, since it feels less hacky to me, but I don't have any strong preference here - care to choose?

I prefer A to B, because:
1. It is simpler, and less code changes.
2. For larger devices, we use long `JMP`, and let the linker to do the branch relaxation.

What's more,
1. we only add this new TargetFeature to devices with larger flash but without long `JMP`. Devices with flash size less than 4KB or with long `JMP` are unnecessary.
2. Generally I prefer `force relocation`, but if the devices have a unique flash size (for example 8KB), I also accept hard coding it into `adjustRelativeBranch` and in your 'Alternatives'.

**Comment by Patryk27** [comment]
Okie, ready! I've adjusted the description above as well. cc @benshi001

**Comment by Patryk27** [comment]
Rebased due to a conflict over https://github.com/llvm/llvm-project/commit/806761a7629df268c8aed49657aeccffa6bca449.

**Comment by benshi001** [comment]
> Rebased due to a conflict over [806761a](https://github.com/llvm/llvm-project/commit/806761a7629df268c8aed49657aeccffa6bca449).

Are you now on the Christmas holiday? What's your opinion on my comments? Do you mind I do the modification based on your code?

**Comment by Patryk27** [comment]
> What's your opinion on my comments? 

Which comments? 👀 

You mentioned using the feature-based approach, which I did implement - I don’t see any more comments 😅

**Comment by benshi001** [comment]
截屏2024-12-24 09 04 57
截屏2024-12-24 09 05 03
截屏2024-12-24 09 05 59

**Comment by benshi001** [comment]
I have posted three review advices as above .

**Review by benshi001 — CHANGES_REQUESTED** [review]

**Comment by Patryk27** [comment]
Cool, thanks! I'll take a look at the comments tomorrow 🙂

File: llvm/lib/Target/AVR/AVRDevices.td [hunk 1]
```diff
@@ -239,7 +251,7 @@ class Device<string Name, Family Fam, ELFArch Arch,
 // in AVR binutils. We do not replicate this.
 def : Device<"avr1", FamilyAVR1, ELFArchAVR1>;
 def : Device<"avr2", FamilyAVR2, ELFArchAVR2>;
-def : Device<"avr25", FamilyAVR25, ELFArchAVR25>;
+def : Device<"avr25", FamilyAVR25, ELFArchAVR25, [FeatureWrappingRjmp]>;
```
[the remark this answers is not in the document]
Comment (Patryk27) [reply llvm/lib/Target/AVR/AVRDevices.td:254]: Sure - fwiw this was synchronized with gcc:

https://github.com/gcc-mirror/gcc/blob/d061c681814917c2208f383bd6069e6ab90db220/gcc/config/avr/avr-mcus.def#L88

File: llvm/test/CodeGen/AVR/branch-relaxation-long-backward.ll [hunk 2]
```diff
@@ -0,0 +1,2078 @@
+; RUN: llc < %s -march=avr -mcpu=avr25 -filetype=obj -o - | llvm-objdump --mcpu=avr25 -dr --no-show-raw-insn --no-leading-addr - | FileCheck --check-prefix=AVR25 %s
+; RUN: llc < %s -march=avr -mcpu=avr3 -filetype=obj -o - | llvm-objdump --mcpu=avr3 -dr --no-show-raw-insn --no-leading-addr - | FileCheck --check-prefix=AVR3 %s
+
```
[the remark this answers is not in the document]
Comment (Patryk27) [reply llvm/test/CodeGen/AVR/branch-relaxation-long-backward.ll:4]: Sure, test extended.

**Comment by Patryk27** [comment]
Ready!

**Review by benshi001 — APPROVED** [review]
LGTM. Thanks.

**Comment by mbuesch** [comment]
Does this fix `attiny26`, too?

I get this
```
error: out of range branch target (expected an integer in the range -4096 to 4095)
```
with the nightly rust compiler on an `attiny26` target, too.

Would this change fix this, even though it doesn't apply `FeatureWrappingRjmp` to `attiny26`?

**Comment by mbuesch** [comment]
I get this behavior with recent rust-nightly in this completely empty project:

[attiny26test.tar.gz](https://github.com/user-attachments/files/18263935/attiny26test.tar.gz)

It does build and work fine with `nightly-2024-03-22`. See `rust-toolchain.toml`.

**Comment by Patryk27** [comment]
Thanks for the reproducer, I'll take a look!

**Comment by MaskRay** [comment]
The PR introduced two tests, which each larger than 81KiB. There are many `call void asm sideeffect "nop", ""()`. You could replace then with a `.space` directive`. See other targets' tests named `branch-relaxation.ll`

**Comment by Patryk27** [comment]
Thanks for the tip, I agree the duplicated `nop`s are unfortunate - I was playing with the `.space` directive, but it didn't work as intended, I'll play around with it again.

File: llvm/lib/Target/AVR/AVRDevices.td [hunk 3]
```diff
@@ -60,6 +60,18 @@ def FeatureSmallStack
                        "The device has an 8-bit "
                        "stack pointer">;
```
Comment (ashutoshmishra52) [inline llvm/lib/Target/AVR/AVRDevices.td:62]: hey

**Comment by ashutoshmishra52** [comment]
xv bx

**Pull request merged**